### PR TITLE
Intuitive Json Detection

### PR DESCRIPTION
**Pull request by gme5078**
So this will improve the way JSONs detect complex JSON files. This allows nested lists of dictionaries to be appropriately detected. It also separates "Data" to have it's own detection so it is detected properly as well. It's a somewhat hacky approach but it works. Please give me more JSON files to try to make the system more robust.

Few questions to consider:
If should all the metadata be put into a metadata property?
Should the intuitive detection default to true or false?

### CODE REVIEW

File: dataprofiler/data_readers/json_data.py [hunk 1]
```diff
@@ -63,21 +64,93 @@ def __init__(self, input_file_path=None, data=None, options=None):
...
+            path = path + "."
+
+        list_of_dict = []
+        if isinstance(json_lines, dict):
+            for key in json_lines:
+                if isinstance(json_lines[key], dict): # If nested dictionary
+                    list_of_dict = list_of_dict + self._find_data(
+                        json_lines[key], path + key)
+                elif isinstance(json_lines[key], list): # If list
+                    if isinstance(json_lines[key][0], dict): # If list of dictionaries
```
Comment (ChrisWallace2020) [inline dataprofiler/data_readers/json_data.py:86]: Should probably have check to make sure its not an empty list, otherwise this index will throw an error.

File: dataprofiler/tests/data_readers/test_json_data.py [hunk 2]
```diff
@@ -135,6 +135,21 @@ def test_len_data(self):
...
+        math = Data(input_file_name, 
+                    options={"intuitive_detection": True})
+
+        self.assertTrue("meta.view.columns.cachedContents.largest" 
+                        in math.data.columns)
+        self.assertEqual(math.data["meta.view.columns.cachedContents.largest"][9]
+                         ,"102188")
+        self.assertTrue("data.22" in math.data.columns)
+        self.assertEqual(math.data["data.22"][167], "77.9")
+
```
Comment (ChrisWallace2020) [inline dataprofiler/tests/data_readers/test_json_data.py:152]: I think adding test cases for `_find_data` would be useful too

**Review by ChrisWallace2020 — CHANGES_REQUESTED** [review]
I think theres a lot of possible improvement for _find_data from a structural standpoint. I think it should be recursive with just 3 cases:
if json_lines is a string, return it
if json_lines is a list, return it
if json_lines is a dict. unpack and recurse

I think type checking what json_lines is past 1 level of nesting is unnecessary and redundant and would look cleaner if broken into these 3 cases more clearly.

File: dataprofiler/data_readers/json_data.py [hunk 3]
```diff
@@ -54,6 +54,7 @@ def __init__(self, input_file_path=None, data=None, options=None):
         self._data_formats["records"] = self._get_data_as_records
         self._data_formats["json"] = self._get_data_as_json
         self._selected_data_format = options.get("data_format", "dataframe")
+        self._intuitive_detection = options.get("intuitive_detection", False)
```
Comment (AnhTruong) [inline dataprofiler/data_readers/json_data.py:57]: the name is somewhat confusing to me. Can we have a more intuitive name?

File: dataprofiler/data_readers/json_data.py [hunk 4]
```diff
@@ -63,21 +64,93 @@ def __init__(self, input_file_path=None, data=None, options=None):
...
+                        json_lines[key], path + key)
+                elif isinstance(json_lines[key], list): # If list
+                    if isinstance(json_lines[key][0], dict): # If list of dictionaries
+                        for item in json_lines[key]:
+                            list_of_dict = list_of_dict + self._find_data(item, path+key)
+                    else: # if non-dict list
+                        list_of_dict = list_of_dict + [{path+key: json_lines[key]}]
+                else: # If just a dictionary of non-dict non-list values
+                    list_of_dict = list_of_dict + [{path+key: json_lines[key]}]
+        else:
```
Comment (AnhTruong) [inline dataprofiler/data_readers/json_data.py:93]: can `json_lines ` be a string?

File: dataprofiler/data_readers/json_data.py [hunk 5]
```diff
@@ -63,21 +64,93 @@ def __init__(self, input_file_path=None, data=None, options=None):
     def selected_keys(self):
         return self._selected_keys
 
+    def _find_data(self, json_lines, path=""):
+        """
+        Finds all the column headers/data in a Json and returns them as a 
+        list.
+
+        :param json_lines: json format list of dicts or dict
+        :type json_lines: list or string or dict
```
Comment (AnhTruong) [inline dataprofiler/data_readers/json_data.py:73]: please add descriptions for `path`

File: dataprofiler/tests/data_readers/test_json_data.py (hunk 2 above)
Comment (AnhTruong) [reply dataprofiler/tests/data_readers/test_json_data.py:152]: agree

File: dataprofiler/data_readers/json_data.py [hunk 6]
```diff
@@ -63,21 +64,93 @@ def __init__(self, input_file_path=None, data=None, options=None):
...
+                            if list(item.keys())[0] not in dict_items:
+                                dict_items[list(item.keys())[0]] = item[
+                                    list(item.keys())[0]]
+                                found = True
+                                break
+                        if found == False:
+                            coalesced_list_of_additions.append(item)
+
+            data, original_df_dtypes = data_utils.json_to_dataframe(
+                json_lines=coalesced_list_of_additions,
```
Comment (AnhTruong) [inline dataprofiler/data_readers/json_data.py:139]: if `json_lines ` is not dict,  `coalesced_list_of_additions ` does not exist. Can you add a fix on this?

**Review by AnhTruong — COMMENTED** [review]
some initial reviews

File: dataprofiler/data_readers/json_data.py [hunk 7]
```diff
@@ -63,21 +64,93 @@ def __init__(self, input_file_path=None, data=None, options=None):
...
-            json_lines = [json_lines]
-        data, original_df_dtypes = data_utils.json_to_dataframe(
-            json_lines=json_lines,
-            selected_columns=self.selected_keys,
-            read_in_string=False
-        )
+        if self._intuitive_detection:
+            bulk_data = None
+            if isinstance(json_lines, dict):
+                if "data" in json_lines.keys():
```
Comment (lettergram) [inline dataprofiler/data_readers/json_data.py:109]: We might want to add a comment here or check several terms (`payload` for instance)

**Review by lettergram — CHANGES_REQUESTED** [review]
I think we need 5-10 tests for this feature (can be small), to ensure it's working as expected.

**Comment by ChrisWallace2020** [comment]
Also, should probably tie this to issue #105 so that there is a link between the issue and the PR.

File: dataprofiler/data_readers/json_data.py [hunk 8]
```diff
@@ -53,17 +55,37 @@ def __init__(self, input_file_path=None, data=None, options=None):
...
+    @property
+    def data_and_metadata(self):
+        """
+        Returns a data frame that joins the data and the metadata.
+        """
+        data = self.data
+        if self._metadata is not None:
+            data = [self._metadata, data]
+            data = pd.concat(data, axis=1)
+        return data
```
Comment (JGSweets) [inline dataprofiler/data_readers/json_data.py:88]: This is an opinion, but i think returning as:
`{'data': self.data, 'metadata': self.metadata}`
might make more sense as it may more accurately represent original state

Comment (JGSweets) [reply dataprofiler/data_readers/json_data.py:88]: Also, metadata key or data key (payload) etc we should think about adding as an option in a future PR

Comment (JGSweets) [reply dataprofiler/data_readers/json_data.py:88]: i see data is an option, metadata could be as well?

File: dataprofiler/data_readers/json_data.py [hunk 9]
```diff
@@ -136,10 +224,23 @@ def _get_data_as_records(self, data):
         return super(JSONData, self)._get_data_as_records(data)
 
     def _get_data_as_json(self, data):
+        data = self._get_data_as_df(data)
         data = data.to_json(orient="records")
         char_per_line = min(len(data), self.SAMPLES_PER_LINE_DEFAULT)
         return list(map(''.join, zip(*[iter(data)] * char_per_line)))
 
+
+    def _get_data_as_df(self, data):
```
Comment (JGSweets) [inline dataprofiler/data_readers/json_data.py:233]: should prove move above the funcs calling it

Comment (JGSweets) [reply dataprofiler/data_readers/json_data.py:233]: Should we add description for these?

File: dataprofiler/data_readers/json_data.py [hunk 10]
```diff
@@ -53,81 +55,167 @@ def __init__(self, input_file_path=None, data=None, options=None):
         #  _selected_keys: keys being selected from the entire dataset
         self._data_formats["records"] = self._get_data_as_records
         self._data_formats["json"] = self._get_data_as_json
+        self._data_formats["data_stream"] = self._get_data_as_data_stream
```
Comment (JGSweets) [inline dataprofiler/data_readers/json_data.py:58]: will need to update the ___init__ descrip above

File: dataprofiler/data_readers/json_data.py [hunk 11]
```diff
@@ -53,81 +55,167 @@ def __init__(self, input_file_path=None, data=None, options=None):
...
+        lists of dictionaries and a key value for a payload.
+
         :param json_lines: json format list of dicts or dict
         :return:
         """
+        bulk_data = None
         if isinstance(json_lines, dict):
-            json_lines = [json_lines]
+            if self._data_key in json_lines.keys():
+                print("HERERER")
```
Comment (AnhTruong) [inline dataprofiler/data_readers/json_data.py:132]: remove?

File: dataprofiler/data_readers/json_data.py [hunk 12]
```diff
@@ -53,81 +55,167 @@ def __init__(self, input_file_path=None, data=None, options=None):
...
+                            dict_items[list(item.keys())[0]] = item[
+                                list(item.keys())[0]]
+                            found = True
+                            break
+                    if found == False:
+                        coalesced_list_of_additions.append(item)
+
         data, original_df_dtypes = data_utils.json_to_dataframe(
-            json_lines=json_lines,
+            json_lines=coalesced_list_of_additions,
```
Comment (AnhTruong) [inline dataprofiler/data_readers/json_data.py:164]: if json_lines is not dict, what is the value of `coalesced_list_of_additions `?

**Comment by JGSweets** [comment]
@gme5078 update branch

File: dataprofiler/data_readers/data_utils.py [hunk 13]
```diff
@@ -180,6 +180,59 @@ def read_json_df(data_generator, selected_columns=None, read_in_string=False):
...
+    :param data_generator: The generator you want to read.
+    :type data_generator: generator
+    :param selected_columns: a list of keys to be processed
+    :type selected_columns: list(str)
+    :param read_in_string: if True, all the values in dataframe will be
+        converted to string
+    :type read_in_string: bool
+    :return: returns an iterator that returns a chunk of file as dataframe in
+        each call as well as original dtypes of the dataframe columns.
+    :rtype: typle(Iterator(pd.DataFrame), pd.Series(dtypes)
```
Comment (ChrisWallace2020) [inline dataprofiler/data_readers/data_utils.py:206]: Is this supposed to say tuple? If it is then its a typo, but also isn't it incorrect since you are returning a list? Also the parenthesis don't close.

File: dataprofiler/data_readers/avro_data.py [hunk 14]
```diff
@@ -47,13 +47,14 @@ def _load_data_from_file(self, input_file_path):
...
-            self._original_df_dtypes = original_df_dtypes
-            return df
+            return lines
+            #df, original_df_dtypes = data_utils.json_to_dataframe(
+            #        json_lines=lines,
+            #        selected_columns=self.selected_keys,
+            #        read_in_string=True
+            #)
+            #self._original_df_dtypes = original_df_dtypes
+            #return df
```
Comment (lettergram) [inline dataprofiler/data_readers/avro_data.py:57]: remove? Also doesn't this potentially create issues because the types aren't discovered?

**Review by lettergram — COMMENTED** [review]
Plz add a couple more JSON examples

File: README.md [hunk 15]
```diff
@@ -615,8 +615,12 @@ may also be specified using the options dict parameter.
 
 Possible `options`:
 
-* data_format - must be a string, choices: "dataframe", "records", "json"
+* data_format - must be a string, choices: "dataframe", "records", "json", "data_stream"
```
Comment (lettergram) [inline README.md:618]: I'm not sure `data_stream` is clear, `payload` or something like that?

File: README.md [hunk 16]
```diff
@@ -615,8 +615,12 @@ may also be specified using the options dict parameter.
...
+  * "data stream" refers to a JSON structure typically found in data streams that contain
```
Comment (lettergram) [inline README.md:619]: IMO this needs a brief example, i.e. { "data": [ columns ], "response": 200}

**Review by lettergram — COMMENTED** [review]
Examples:

* Deeply nested JSON objects
* Switching the data keys (to `hits`): https://hn.algolia.com/api/v1/search_by_date?tags=comment&numericFilters=created_at_i%3E20200103
* No keys provided, see if meta data is still obtained

File: dataprofiler/tests/data_readers/test_json_data.py [hunk 17]
```diff
@@ -136,6 +136,39 @@ def test_len_data(self):
...
+            
+    def test_data_stream_format(self):
+        test_dir = os.path.join(test_root_path, 'data')
+        input_file_name = os.path.join(test_dir, 'json/math.json')
+
+        math = Data(input_file_name, 
+                    options={"data_format": "data_stream"})
+
+        self.assertTrue("meta.view.columns.cachedContents.largest" 
+                        in math.data_and_metadata.columns)
```
Comment (ChrisWallace2020) [inline dataprofiler/tests/data_readers/test_json_data.py:166]: You can use "self.assertIn" for this case and the one below

File: README.md [hunk 18]
```diff
@@ -615,8 +615,12 @@ may also be specified using the options dict parameter.
 
 Possible `options`:
 
-* data_format - must be a string, choices: "dataframe", "records", "json"
+* data_format - must be a string, choices: "dataframe", "records", "json", "flattened_dataframe"
+  * "flattened_dataframe" is best used for JSON structure typically found in data streams that contain
+nested lists of dictionaries and a payload. For example: `{"data": [ columns ], "response": 200}`
 * selected_keys - columns being selected from the entire dataset, must be a list `["column 1", "ssn"]`
+* payload_key - The dictionary key for the payload of the JSON, typically called "data"
+or "payload". Defaults to "data".
```
Comment (lettergram) [inline README.md:623]: Should it be 'payload_keys' and set it as a list, e.g. should already be `["data", "response"]` at the very least

File: dataprofiler/data_readers/data_utils.py [hunk 19]
```diff
@@ -180,6 +180,59 @@ def read_json_df(data_generator, selected_columns=None, read_in_string=False):
...
+                ignore_dicts=True
+            )
+            lines.append(obj)
+        except ValueError:
+            pass
+            # To ignore malformatted lines.
+        k += 1
+    if not lines and k:
+        raise ValueError('No JSON data could be read from these data.')
+    return lines
```
Comment (lettergram) [inline dataprofiler/data_readers/data_utils.py:233]: yield?

File: dataprofiler/data_readers/data_utils.py [hunk 20]
```diff
@@ -180,6 +180,59 @@ def read_json_df(data_generator, selected_columns=None, read_in_string=False):
     return json_to_dataframe(lines, selected_columns, read_in_string)
 
 
+def read_json(data_generator, selected_columns=None, read_in_string=False):
+    """
+    This function returns an iterator that returns a chunk of data
```
Comment (ChrisWallace2020) [inline dataprofiler/data_readers/data_utils.py:185]: Doesn't this function return a list, not an iterator? There are no yield statements anywhere. You claim its an iterator below in the :returns: section too, but is this still accurate?

File: dataprofiler/tests/data_readers/test_json_data.py (hunk 2 above)
Comment (ChrisWallace2020) [reply dataprofiler/tests/data_readers/test_json_data.py:182]: You resolved the conversation but haven't added tests for `_find_data`, do you think they aren't necessary?

File: dataprofiler/data_readers/json_data.py [hunk 21]
```diff
@@ -23,14 +25,17 @@ def __init__(self, input_file_path=None, data=None, options=None):
         Possible Options::
         
             options = dict(
-                data_format= type: str, choices: "dataframe", "records", "json"
+                data_format= type: str, choices: "dataframe", "records", "json",
+                 "data_stream"
                 selected_keys= type: list(str)
+                data_key= type: str
```
Comment (ChrisWallace2020) [inline dataprofiler/data_readers/json_data.py:31]: These are out of date with updates in README

File: dataprofiler/data_readers/json_data.py [hunk 22]
```diff
@@ -23,14 +25,17 @@ def __init__(self, input_file_path=None, data=None, options=None):
...
             )
 
         
         data_format: user selected format in which to return data
         can only be of specified types
         selected_keys: keys being selected from the entire dataset
+        data_key: dictionary key that determines the payload
```
Comment (ChrisWallace2020) [inline dataprofiler/data_readers/json_data.py:38]: Is this payload_keys now?

File: dataprofiler/data_readers/json_data.py [hunk 23]
```diff
@@ -51,83 +56,201 @@ def __init__(self, input_file_path=None, data=None, options=None):
...
         #  _selected_keys: keys being selected from the entire dataset
+        #  _payload_keys: dictionary key that determines the payload
+
         self._data_formats["records"] = self._get_data_as_records
         self._data_formats["json"] = self._get_data_as_json
-        self._selected_data_format = options.get("data_format", "dataframe")
+        self._data_formats["flattened_dataframe"] = self._get_data_as_flattened_dataframe
+        self._selected_data_format = options.get("data_format", "flattened_dataframe")
+        self._payload_keys = options.get("payload_keys", ["data", "payload", "response"])
+        if not isinstance(self._payload_keys, list):
```
Comment (ChrisWallace2020) [inline dataprofiler/data_readers/json_data.py:66]: Should you make sure its a string first or can it be float/int too?

File: README.md (hunk 15 above)
Comment (gme5078) [reply README.md]: Switching to flattened_dataframe because that is what it is

File: dataprofiler/data_readers/data_utils.py (hunk 19 above)
Comment (gme5078) [reply dataprofiler/data_readers/data_utils.py:233]: Fixed the comments, it is not an iterator. Lot of vestigial descriptions from where I copied this.

File: dataprofiler/data_readers/data_utils.py (hunk 20 above)
Comment (gme5078) [reply dataprofiler/data_readers/data_utils.py:185]: Fixed the comments, it is not an iterator. Lot of vestigial descriptions from where I copied this.

File: dataprofiler/data_readers/json_data.py (hunk 23 above)
Comment (gme5078) [reply dataprofiler/data_readers/json_data.py:66]: Technically json only has string keys, but dictionaries don't need to be. So maybe the data is sent in as a dict to be read by JSON data with integer keys or maybe someone will send in an integer key but actually mean string. I'm not sure which use case would ever happen so I'm not sure what to do here, but I dont think it really matters

File: dataprofiler/data_readers/json_data.py [hunk 24]
```diff
@@ -23,14 +25,17 @@ def __init__(self, input_file_path=None, data=None, options=None):
         Possible Options::
         
             options = dict(
-                data_format= type: str, choices: "dataframe", "records", "json"
+                data_format= type: str, choices: "dataframe", "records", "json",
+                 "data_stream"
```
Comment (lettergram) [inline dataprofiler/data_readers/json_data.py:29]: don't we not have "data_stream" any more?

File: dataprofiler/tests/data_readers/test_json_data.py [hunk 25]
```diff
@@ -136,6 +136,110 @@ def test_len_data(self):
                              data.length,
                              msg=input_file['path'])
 
+    def test_data_stream_format_with_no_payload(self):
```
Comment (lettergram) [inline dataprofiler/tests/data_readers/test_json_data.py:139]: data_stream

File: dataprofiler/tests/data_readers/test_json_data.py [hunk 26]
```diff
@@ -136,6 +136,110 @@ def test_len_data(self):
...
+
+    def test_list_of_dictionaries_in_payload_format(self):
+        test_dir = os.path.join(test_root_path, 'data')
+        input_file_name = os.path.join(test_dir, 'json/iris-utf-8.json')
+
+        simple = Data(input_file_name)
+        self.assertEqual(6, len(simple.data.columns))
+        self.assertEqual(150, len(simple.data))
+        
+    def test_data_stream_format(self):
```
Comment (lettergram) [inline dataprofiler/tests/data_readers/test_json_data.py:191]: data_stream

File: dataprofiler/data_readers/json_data.py [hunk 27]
```diff
@@ -51,83 +56,201 @@ def __init__(self, input_file_path=None, data=None, options=None):
...
         :param json_lines: json format list of dicts or dict
         :return:
         """
+        if isinstance(json_lines, pd.DataFrame):
+            return json_lines
+        payload_data = None
         if isinstance(json_lines, dict):
-            json_lines = [json_lines]
+            # Glean Payload Data
+            found_payload_key = None
```
Comment (lettergram) [inline dataprofiler/data_readers/json_data.py:162]: What if multiple payloads found?

Should probably add a test and clarify what'll happen here.

File: dataprofiler/data_readers/json_data.py [hunk 28]
```diff
@@ -51,83 +56,201 @@ def __init__(self, input_file_path=None, data=None, options=None):
...
+                        json_lines=payload_data,
+                        selected_columns=self.selected_keys,
+                        read_in_string=False
+                    )
+                    for column in payload_data.columns:
+                        payload_data.rename(
+                            columns={column: payload_key + self._key_separator + str(column)},
+                            inplace=True)
+                    found_payload_key = payload_key
+                    break
```
Comment (lettergram) [inline dataprofiler/data_readers/json_data.py:179]: I see the break on if a payload is found, does the other payload (if two are found) become metadata. Could select largest one?

File: dataprofiler/data_readers/json_data.py [hunk 29]
```diff
@@ -51,83 +56,201 @@ def __init__(self, input_file_path=None, data=None, options=None):
...
-                    self._load_data_from_json_format(json_lines)
+                data = json.load(input_file)
             except (json.JSONDecodeError, UnicodeDecodeError):
                 input_file.seek(0)
-                data_generator = data_utils.generator_on_file(input_file)
-                data, original_df_dtypes = data_utils.read_json_df(
-                    data_generator=data_generator,
+                data = data_utils.generator_on_file(input_file)
+                data = data_utils.read_json(
+                    data_generator=data,
```
Comment (lettergram) [inline dataprofiler/data_readers/json_data.py:238]: Couldn't this fail as well?

File: dataprofiler/data_readers/json_data.py [hunk 30]
```diff
@@ -51,83 +56,201 @@ def __init__(self, input_file_path=None, data=None, options=None):
...
+
+        :param list_of_dicts: the list of dictionaries to be coalesced
+        :type list_of_dicts: list(dict)
+        :return: Coalesced list of dictionaries
+        """
+        coalesced_list_of_dicts = [{}]
+        for item in list_of_dicts:
+            found = False
+            for dict_items in coalesced_list_of_dicts:
+                if list(item.keys())[0] not in dict_items:
```
Comment (ChrisWallace2020) [inline dataprofiler/data_readers/json_data.py:141]: Maybe leave a comment that you expect these dictionaries to only have 1 item

File: dataprofiler/data_readers/json_data.py [hunk 31]
```diff
@@ -51,83 +56,201 @@ def __init__(self, input_file_path=None, data=None, options=None):
...
+        return coalesced_list_of_dicts
+
+    def _get_data_as_flattened_dataframe(self, json_lines):
+        """
+        Loads the data when in a JSON format from a data stream (nested
+        lists of dictionaries and a key value for a payload.
+
         :param json_lines: json format list of dicts or dict
         :return:
         """
```
Comment (ChrisWallace2020) [inline dataprofiler/data_readers/json_data.py:156]: Should update this docstring to describe what is returned

File: dataprofiler/data_readers/json_data.py [hunk 32]
```diff
@@ -23,14 +25,17 @@ def __init__(self, input_file_path=None, data=None, options=None):
         Possible Options::
         
             options = dict(
-                data_format= type: str, choices: "dataframe", "records", "json"
+                data_format= type: str, choices: "dataframe", "records", "json",
+                 "data_stream"
                 selected_keys= type: list(str)
+                payload_keys= type: str
```
Comment (ChrisWallace2020) [inline dataprofiler/data_readers/json_data.py:31]: Isn't this a list of strs?

File: dataprofiler/data_readers/json_data.py (hunk 24 above)
Comment (gme5078) [reply dataprofiler/data_readers/json_data.py:29]: swapped it for flattened dataframe, somewhat more apt

File: dataprofiler/data_readers/json_data.py (hunk 27 above)
Comment (gme5078) [reply dataprofiler/data_readers/json_data.py:162]: Just takes the first payload it finds

File: dataprofiler/data_readers/json_data.py (hunk 29 above)
Comment (gme5078) [reply dataprofiler/data_readers/json_data.py:238]: Maybe, but if it fails there, it should error out

File: dataprofiler/data_readers/json_data.py (hunk 32 above)
Comment (gme5078) [reply dataprofiler/data_readers/json_data.py:31]: both, adding

File: dataprofiler/data_readers/json_data.py (hunk 28 above)
Comment (lettergram) [reply dataprofiler/data_readers/json_data.py:181]: I guess most rows 🤷‍♂️, if they match they should probably create a single dataset?

File: dataprofiler/data_readers/json_data.py [hunk 33]
```diff
@@ -51,83 +56,210 @@ def __init__(self, input_file_path=None, data=None, options=None):
...
-                    self._load_data_from_json_format(json_lines)
+                data = json.load(input_file)
             except (json.JSONDecodeError, UnicodeDecodeError):
                 input_file.seek(0)
-                data_generator = data_utils.generator_on_file(input_file)
-                data, original_df_dtypes = data_utils.read_json_df(
-                    data_generator=data_generator,
+                data = data_utils.generator_on_file(input_file)
+                data = data_utils.read_json(
+                    data_generator=data,
```
Comment (lettergram) [inline dataprofiler/data_readers/json_data.py:247]: wouldn't use the same name 'data' - maybe `file_generator` or something

**Review by lettergram — APPROVED** [review]
I'm a little concerned the code for all the various edge cases for this are not here, but they are likely unknown (I couldn't think of any more to test off hand). I think we'll find more after some the utilization.

**Review by ChrisWallace2020 — APPROVED** [review]
